### PR TITLE
Add options for minimal links and auto-prefixing http protocol

### PR DESCRIPTION
**Pull request by SDRACK**
Two options added: `autoPrefixHttpProtocol` and `minimalLinks`.

`minimalLinks` slims down the link overlay to use only `text` and `url`.

`autoPrefixHttpProtocol` adjust links without a protocol. Those with a protocol, anchors and relative links are left alone. Anything else if prepended with `http://` or `mailto:` as appropriate. 

Both new options default to `false`.

N.B. ES6 methods used for string interpolation. Can adjust if required.

### CODE REVIEW

File: src/trumbowyg.js [hunk 1]
```diff
@@ -1232,6 +1248,15 @@ Object.defineProperty(jQuery.trumbowyg, 'defaultOptions', {
...
             });
         },
+        prependHttpProtocol: function(url) {
+            var t = this;
+            if(!t.o.autoPrefixHttpProtocol) { return url; }
+
+            const VALID_LINK_PREFIX = /^([a-z][-+.a-z0-9]*:|\/|#)/i;
+            if(VALID_LINK_PREFIX.test(url)) { return url; }
+
+            return url.includes("@") ? `mailto:${url}` : `http://${url}`;
```
Comment (Alex-D) [inline src/trumbowyg.js]: That's wrong, you can use @ in HTTP url to put some user/password:

`[//[user[:password]@]host[:port]][/path][?query][#fragment]`

**Comment by Alex-D** [comment]
Do not use ES6 since I did not use babel and want to be ES5-compatible.

**Comment by SDRACK** [comment]
Cheers for the feedback - I'll update the ES6 bits and implement a regex to distinguish email addresses / urls.

Be back with you soon 👍

**Comment by forelabs** [comment]
I suggest to rename it to autoPrefixProtocol and have an urlProtocol option which defaults to https, but can be set to e.g. Http. Https will be the new Standart since chrome will also mark websites as untrusted soon when they use http.

**Comment by SDRACK** [comment]
Thanks, sounds sensible. I've updated the option name. Under the new setup, users can provide the following options to `urlProtocol`:

- `false` (default)
- `true`, `'https'` or `'https://'` => `'https://'`
- `'http'` or `'http://'` => `'http://'`

Anything else will be ignored.

Let me know what you think - appreciate your work on this!

File: src/trumbowyg.js [hunk 2]
```diff
@@ -472,6 +474,23 @@ Object.defineProperty(jQuery.trumbowyg, 'defaultOptions', {
...
 
+        setupUrlPrefix: function() {
+          var t = this,
+              protocol = t.o.urlProtocol;
+
+          if(!protocol) { return; }
+
+          if(typeof(protocol) === 'boolean') {
+              return 'https://'
+          } else if(typeof(protocol) === 'string') {
```
Comment (Alex-D) [inline src/trumbowyg.js]: Don't need else as we always manage all others cases with the previous if

File: src/trumbowyg.js [hunk 3]
```diff
@@ -472,6 +474,23 @@ Object.defineProperty(jQuery.trumbowyg, 'defaultOptions', {
             this.btnsDef[btnName] = btnDef;
         },
 
+        setupUrlPrefix: function() {
+          var t = this,
+              protocol = t.o.urlProtocol;
+
+          if(!protocol) { return; }
+
+          if(typeof(protocol) === 'boolean') {
```
Comment (Alex-D) [inline src/trumbowyg.js]: `if (typeof protocol !== 'string') {`

File: src/trumbowyg.js [hunk 4]
```diff
@@ -1199,31 +1222,43 @@ Object.defineProperty(jQuery.trumbowyg, 'defaultOptions', {
...
-                    value: title
-                },
                 text: {
                     label: t.lang.text,
-                    value: text
-                },
-                target: {
-                    label: t.lang.target,
-                    value: target
+                    value: new XMLSerializer().serializeToString(documentSelection.getRangeAt(0).cloneContents())
```
Comment (forelabs) [inline src/trumbowyg.js]: this one is out of date and should stay `text`

File: src/trumbowyg.js [hunk 5]
```diff
@@ -480,16 +480,10 @@ Object.defineProperty(jQuery.trumbowyg, 'defaultOptions', {
...
-              if(/^https?:\/\/$/.test(protocol)) { return protocol; }
-
-              if(/^https?$/.test(protocol)) {
-                  return protocol + '://'
-              }
-          }
-        }
+          if(typeof(protocol) !== 'string') { return 'https://'; }
+          if(/^https?:\/\/$/.test(protocol)) { return protocol; }
+          if(/^https?$/.test(protocol)) { return protocol + '://'; }
```
Comment (Alex-D) [inline src/trumbowyg.js]: That's a lot of code to just get `https://` or `http://` :/

Comment (Alex-D) [reply src/trumbowyg.js]: What about others protocols? If I want put some ftp:// or whatever?

Comment (SDRACK) [reply src/trumbowyg.js]: Thanks @Alex-D. 

Would you prefer to allow any string value in the option, and append `://` if it's missing?

So:

- a `true` value defaults to `https://`
- a string without the correct suffix has `://` appended to it
- a string with the correct suffix is prepended as is

Comment (Alex-D) [reply src/trumbowyg.js]: yep I think it could be cool :)

**Comment by SDRACK** [comment]
Morning @Alex-D - that's the `setupUrlPrefix` function shortened and set to allow any string as said prefix. For easy reference, here's the updated code:

```
        setupUrlPrefix: function() {
            var protocol = this.o.urlProtocol;
            if(!protocol) { return; }

            if(typeof(protocol) !== 'string') { return 'https://'; }
            return /:\/\/$/.test(protocol) ? protocol : protocol + '://';
        }
```

How does that look to you?

**Comment by Alex-D** [comment]
LGTM :)

Could you update the doc about this new option? It's somewhere here: 

- https://github.com/Alex-D/Trumbowyg/blob/develop/docs/documentation/index.html#L64
- https://github.com/Alex-D/Trumbowyg/blob/develop/docs/documentation/index.html#L463

Thanks :)

**Comment by SDRACK** [comment]
Fantastic - thanks @Alex-D. Really appreciate all your work with me on this. 

Let me know what you think of the updated docs and I'll edit if needed.

One thing to note: I've put `2.9.4` as the version added as I'm not sure of your process for iterating this. Let me know if that needs tweaking.

**Comment by SDRACK** [comment]
@Alex-D - please don't consider this a chase, but I'm wondering what your process is for introducing new features such as these, and when you're next likely to do so?

Hope I'm not being a pain asking this :)

**Comment by Alex-D** [comment]
Just taking time to run this on my computer, then merge other MRs, release a new version, test it, etc.

So I need to take couple of hours ^^'

Sorry if it's long...

**Comment by SDRACK** [comment]
Thanks, that clears things up perfectly. It was more a case of establishing whether this would be deployed shortly or as part of a larger release way down the line. So that answers my question perfectly.

Thanks again for your work on this.

**Comment by Alex-D** [comment]
Thanks! :)

**Comment by jorrit** [comment]
This PR (and release 2.10.0) broke ES5 browsers/minifiers because of the use of `const`.

See #764.

**Pull request merged**